tla-bmcmt/src/main/scala/at/forsyte/apalache/tla/bmcmt/ModelChecker.scala

```diff
@@ -506,22 +510,38 @@ class ModelChecker(typeFinder: TypeFinder[CellT],
     }
   }
 
-  private def dumpCounterexample(): String = {
-    val filename = "counterexample.txt"
+  private def dumpCounterexample(notInv: TlaEx): String = {
+    // Assumption: stack cannot be empty
+    val filename = if(tlc) "MC.out" else "counterexample.tla"
     val writer = new PrintWriter(new FileWriter(filename, false))
+
+    val nextStates = new ListBuffer[NextState]()
     for (((state, oracle), i) <- stack.reverse.zipWithIndex) {
       val decoder = new SymbStateDecoder(solverContext, rewriter)
       val transition = oracle.evalPosition(solverContext, state)
-      writer.println(s"State $i (from transition $transition):")
-      writer.println("--------")
       val binding = decoder.decodeStateVariables(state)
-      val args = binding.toList.sortBy(_._1).flatMap(p => List(ValEx(TlaStr(p._1)), p._2))
-      if (args.nonEmpty) {
-        val record = OperEx(TlaFunOper.enum, args :_*)
-        new PrettyWriter(writer).write(record)
-      }
-      writer.println("\n========\n")
+      nextStates += ((transition.toString, binding))
+    }
```

Comment:
Nice! Shall we actually produce a `TlaModule` here instead of lists?

Reply:
We can fix the names of the declarations for the writers. E.g., `Init` for initialization, `ConstInit` for parameter initialization, `State<no>` for the states in the counterexample, `InvariantViolation` for the invariant negation.

Reply:
In that respect, FWIW, I think it's better to keep specialized methods with specific signatures: packing everything into a TlaModule will introduce unnecessary code for packing/unpacking, and won't bring any real value. Let's maybe keep it as is for now, and we'll refactor later if needed

Reply:
Well, the value is that the counterexample printers become just code printers. The TLC format is somewhat an exception here, as it contains source information.